### ISSUE TITLE
Deploying multiple bridge-domain services fails randomly

### ISSUE THREAD

**Issue by U24543426** [networkservicemesh/networkservicemesh#2208]
## Expected Behavior

When deploying multiple bridge services:
(or a single service with multiple endpoints with different ip subnets, same result)
```
(n1) -- (br1) -- (n2)     (n3) -- (br2) -- (n4)     (n5) -- (br3) -- (n6)
br1: 172.16.0.0/24
br2: 172.16.1.0/24
br3: 172.16.2.0/24
...
(similar results with ip6)
```
Nodes on the same bridge subnet should ping 

## Failure Information (for bugs)

First bridge service works fine, n1 and n2 pings
Then adding the second bridge service also works fine, n3 and n4 pings
However, adding the third and the subsequent bridges would not ping
No outstanding errors during service or bridge endpoint creation 

If you create all the components sequentially with rapid succession (without pause), then none of the services work.

Tried two different ways:
First with a single service definition with different labels
Then with multiple service definitions 
Same result in both cases, only two subnets would work, the rest would not.

## Steps to Reproduce

Nodes are NSCs with initContainers
Bridges are NSEs with networkservicemesh/bridge-domain-bridge
Forwarder vpp

## Context

Details same as in issue #2204 
Ubuntu 18.04
Docker 19.03
Kubernetes 1.19
Kubeadm, single node
Nodes are NSCs with initContainers
Bridges are NSE with networkservicemesh/bridge-domain-bridge
Forwarder vpp
NSM master from Oct 8

## Failure Logs

Service.yml
[service.yml.txt](https://github.com/networkservicemesh/networkservicemesh/files/5365346/service.yml.txt)

Node.yml
[node.yml.txt](https://github.com/networkservicemesh/networkservicemesh/files/5365349/node.yml.txt)

Bridge.yml
[bridge.yml.txt](https://github.com/networkservicemesh/networkservicemesh/files/5365351/bridge.yml.txt)

**Comment by U24543426** [comment]
Getting random fail/success results in different trials. Creating clients manually one at a time seem to give the best success rate.

Eleven-node, 14-bridge random topology generated with the pattern described above (node--bridge--node--)

In one trial, services and bridge-NSEs are created at once, then the clients are created manually one at a time.
In that case 2 bridges work both-directions from each connected node. 
Two bridges work partially, only in one direction from the node in the center.
The remaining 10 bridges fail.

In another trial, all component are created at once (bridges, NSEs, NSCs).
In that case, only one bridge works.
All other 13 bridges fail (nodes would not ping)

When the above trial is repeated, all bridges failed.

In another run created all the services and the clients at once, then each NSE one at a time manually. All but one bridge failed again. 

The bridges that fail and succeed seem to be random.

**Comment by U24543426** [comment]
We did another experiment while trying to find the root cause of the problem ...
(with no luck)

Cross connectivity of the bridge-domains is not necessary to reproduce the problem.
The problem happens when you just create multiple, isolated service instances:

    (n1)---(b)---(n2)    (n3)---(b)---(n4)    (n5)---(b)---(n6)  ... 

When you have 6 of these bridge domains, if you create the clients one-at-a-time, then you end up with only 2 of them working. The first client pair that is defined always pings. The second pair that works is usually random.

If you create everything with a script, right one after the other without any pause in between, then none of the services work--nothing pings.

Every service name and every label in each domain is unique, there is no naming overlap.

If this is not related to multiple instance of the bridge-domain service, then you should also be seeing similar connectivity problems with multiple service deployments when everything is running on a single node, k8s on bare-metal.

**Comment by edwarnicke** [comment]
@haiodo @denis-tingajkin Could we have someone take a look at this?

**Comment by denis-tingajkin** [comment]
@xzfc Could you take look at this?

**Comment by U24543426** [comment]
Can this be related to prefix pools? 
Similar type of connectivity problems happen with the Topology example -- only part of the nodes can get connectivity.
Even though when the Endpoints request /24 address pools, they keep getting /30 pools.
With these service examples, the idea is to give 172.16.1... 172.16.2..., 172.16.3... blocks to different domains.
The /30 leaves just 4 addresses in the pool.
Then maybe the address domain separation of different services are not fully granted and maybe some get served from the same prefix pool?

**Comment by xzfc** [comment]
I can reproduce it, both in Vagrant and in Kind. Starting from 7 or 8 NS clients, the new ones become isolated: they have an interface and an assigned IP, but they are not able to ping anyone else in the same subnet.

It is not affected by the number of bridges, as the issue is reproducible with just one bridge with 8 connected NSC. So, it is reproducible in each of the following scenarios:

```
(n1)---(b1)---(n2)    (n3)---(b2)---(n4)    (n5)---(b3)---(n6)   (n7)---(b4)---(n8)
```

```
(n1)───(b1)───(n2)    (n5)───(b2)───(n6)
(n3)───┘  └───(n4)    (n7)───┘  └───(n8)
```

```
(n1)───(b1)───(n2)
(n3)───┤  ├───(n4)
(n5)───┤  ├───(n6)
(n7)───┘  └───(n8)
```

**Comment by xzfc** [comment]
On the "isolated" pod:

```
/ # ifconfig nsm0
nsm0      Link encap:Ethernet  HWaddr 02:FE:A4:4E:E9:D7  
          inet addr:172.16.0.9  Bcast:172.16.0.255  Mask:255.255.255.0
          inet6 addr: fe80::fe:a4ff:fe4e:e9d7/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:123 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:5682 (5.5 KiB)
```

Note that `RX packets` is `0`. This interface isn't receiving frames. However, it transmits packets just fine, and other pods are able to receive frames from the "isolated" pod. (checked it with `tcpdump` on another pod)

**Comment by xzfc** [comment]
I've reduced the reproducible example to the following issue reported in vpp-agent, which I suppose is the cause of this bridge-domain NSM issue.

https://github.com/ligato/vpp-agent/issues/1758

**Comment by xzfc** [comment]
@U24543426 Could you please check if this issue is reproducible with [this patch](https://github.com/xzfc/networkservicemesh/commit/c24cc13a6871dbdc0edad487e9307338e84293c9) to the VPP config?

---

The issue that I can reproduce is caused by VPP packet buffer pool exhaustion at vpp-forwarder. VPP pre-allocates a certain amount of memory buffers at startup (which can be configured), and these buffers are consumed during interface creation.

Also, this issue affects not only the bridge-domain scenario but also kernel-to-kernel forwardings made by vpp-forwarder. With the default `buffers-per-numa` value, we are limited to 16 kernel-to-kernel interface pairs. The issue doesn't occur if the VETH pairs are used (`FORWARDER_ALLOW_VHOST=false`) instead of TAP interfaces, but the bandwidth is worse (2.44 vs 1.28 Gbits/sec on iperf3 test on my machine).

Diagnostics:
* `vppctl sh buffers`
* `vppctl sh err` contains `no free tx slots` or `buffer allocation failed`
* Kernel interface packet counters (`ip -s link`) are 0

**Comment by U24543426** [comment]
> @U24543426 Could you please check if this issue is reproducible with [this patch](https://github.com/xzfc/networkservicemesh/commit/c24cc13a6871dbdc0edad487e9307338e84293c9) to the VPP config?
> 

Looks pretty good.

I don't have the simple bridge setup right now, but I tried it on a large topology with 70 nodes. Without the patch, at most 3 or 4 nodes got connectivity; with the patch, they all seem to be up and running. (It will be good to see it scale to 100 nodes/NSC/NSE - max number of pods that can be deployed on a single host node.)

Thank you!

**Comment by xzfc** [comment]
Closed in favor of #2218.
